File: py/rmse_function.py
import py.target_function as tf
import numpy as np


class RmseFunction(tf.TargetFunction):

    def __init__(self, target_function):
        self.target_function = target_function

    def fun(self, params, args):
        truth_output = args[:, -1]
        train_input = args[:, :-1]
        res_array = []
        for i in range(train_input.shape[0]):
            input = np.array(train_input[i, :])
            res = self.target_function.fun(params, input)
            res_array.append(res)
        res = np.array(res_array)
        rmse = np.sqrt(np.divide(np.sum(np.power(res - truth_output, 2)), res.shape[0]))
        return rmse

File: src/main/java/jm/app/optimize/RmseFunction.java
package jm.app.optimize;

import jm.app.algebra.AlgebraUtil;
import jm.app.algebra.Matrix;

import java.math.BigDecimal;

public class RmseFunction implements TargetFunction {

    private TargetFunction originalFunction;

    public RmseFunction(TargetFunction originalFunction) {
        this.originalFunction = originalFunction;
    }

    /**
     * Calculate the RMSE of the original function with training samples
     * @param params parameters of the original function
     * @param args arguments include the training input and the truth output, the last column is the truth output
     * @return RMSE of the original function
     */
    @Override
    public BigDecimal fun(Matrix params, Matrix args) {
        Matrix truthOutput = AlgebraUtil.getColumnVector(args, args.getColNum() - 1);
        Matrix trainInput = AlgebraUtil.getSubMatrix(
                args,
                0,
                args.getRowNum() - 1,
                0,
                args.getColNum() - 1 - 1);
        BigDecimal rmse = new BigDecimal(0.0);
        for (int i = 0; i < trainInput.getRowNum(); i++) {
            Matrix input = AlgebraUtil.getRowVector(trainInput, i);
            BigDecimal res = originalFunction.fun(params, input);
            BigDecimal truth = truthOutput.getValue(i, 0);
            rmse = rmse.add(res.subtract(truth).multiply(res.subtract(truth)));
        }
        rmse = rmse.multiply(new BigDecimal(1.0 / trainInput.getRowNum()));
        rmse = new BigDecimal(Math.sqrt(rmse.doubleValue()));
        return rmse;
    }

    public TargetFunction getOriginalFunction() {
        return originalFunction;
    }

    public void setOriginalFunction(TargetFunction originalFunction) {
        this.originalFunction = originalFunction;
    }
}
